File: artifacts/api-server/src/routes/sectors.ts
import { Router, Request, Response } from 'express';
import { NseService } from '../services/nse.service';
import { YahooService } from '../services/yahoo.service';
import { SectorsService } from '../services/sectors.service';

const router = Router();

let sectorsService: SectorsService | null = null;

function getService() {
  if (!sectorsService) {
    const nse = new NseService();
    const yahoo = new YahooService();
    sectorsService = new SectorsService(nse, yahoo);
  }
  return sectorsService;
}

router.get('/', async (req: Request, res: Response) => {
  try {
    const service = getService();
    const data = await service.getAllSectors();
    res.json(data);
  } catch (err: any) {
    req.log.error({ err }, 'Failed to get sectors');
    res.status(500).json({ error: 'Failed to fetch sector data' });
  }
});

router.get('/rotation', async (req: Request, res: Response) => {
  try {
    const service = getService();
    const data = await service.getSectorRotation();
    res.json(data);
  } catch (err: any) {
    req.log.error({ err }, 'Failed to get rotation');
    res.status(500).json({ error: 'Failed to fetch rotation data' });
  }
});

router.get('/:symbol', async (req: Request, res: Response) => {
  try {
    const service = getService();
    const data = await service.getSectorDetail(String(req.params.symbol ?? ""));
    if (!data) {
      res.status(404).json({ error: 'Sector not found' });
      return;
    }
    res.json(data);
  } catch (err: any) {
    req.log.error({ err }, 'Failed to get sector detail');
    res.status(500).json({ error: 'Failed to fetch sector detail' });
  }
});

export default router;

File: artifacts/python-backend/app/routes/sectors.py
from fastapi import APIRouter
from fastapi.responses import JSONResponse
from ..services import registry as svc

router = APIRouter(prefix="/sectors", tags=["sectors"])


async def _get_sectors():
    return await svc.sectors.get_all_sectors()

router.add_api_route("",  _get_sectors, methods=["GET"])
router.add_api_route("/", _get_sectors, methods=["GET"])


@router.get("/rotation")
async def get_rotation():
    return await svc.sectors.get_sector_rotation()


@router.get("/{symbol:path}")
async def get_sector(symbol: str):
    data = await svc.sectors.get_sector_detail(symbol)
    if data is None:
        return JSONResponse(status_code=404, content={"error": f"Sector '{symbol}' not found"})
    return data
